packages/lib/test/computedTree/computedTree.test.ts

```diff
@@ -0,0 +1,518 @@
+    const r = new R({})
+
+    expect(isRoot(r.array)).toBeFalsy()
+    expect(isRoot(r.plainObject)).toBeFalsy()
+    expect(isRoot(r.model)).toBeFalsy()
+  })
+
+  test("computed tree supports isChildOfParent", () => {
+    const r = new R({})
+
+    expect(isChildOfParent(r.array, r)).toBeTruthy()
+    expect(isChildOfParent(r.plainObject, r)).toBeTruthy()
+    expect(isChildOfParent(r.model, r)).toBeTruthy()
+  })
+
+  test("computed tree supports isParentOfChild", () => {
+    const r = new R({})
+
+    expect(isParentOfChild(r, r.array)).toBeTruthy()
+    expect(isParentOfChild(r, r.plainObject)).toBeTruthy()
+    expect(isParentOfChild(r, r.model)).toBeTruthy()
+  })
+
+  // fails because `resolvePath` skips ahead to `<model>.$` when it arrives at a model node
+  test.skip("computed tree supports resolvePath", () => {
+    const r = new R({})
+
+    expect(resolvePath(r, ["array"])).toEqual({ resolved: true, value: r.array })
+    expect(resolvePath(r, ["plainObject"])).toEqual({ resolved: true, value: r.plainObject })
+    expect(resolvePath(r, ["model"])).toEqual({ resolved: true, value: r.model })
+  })
+
+  test("computed tree supports findParent", () => {
+    const r = new R({})
+
+    const isR = (node: unknown) => node instanceof R
+
+    expect(findParent(r.array, isR)).toBe(r)
+    expect(findParent(r.plainObject, isR)).toBe(r)
+    expect(findParent(r.model, isR)).toBe(r)
+  })
+
+  test("computed tree supports findParentPath", () => {
+    const r = new R({})
+
+    const isR = (node: unknown) => node instanceof R
+
+    expect(findParentPath(r.array, isR)).toEqual({ parent: r, path: ["array"] })
+    expect(findParentPath(r.plainObject, isR)).toEqual({ parent: r, path: ["plainObject"] })
+    expect(findParentPath(r.model, isR)).toEqual({ parent: r, path: ["model"] })
+  })
+
+  test("computed tree supports findChildren", () => {
+    const r = new R({})
+
+    // cached when observed
+    autoDispose(
+      reaction(
+        () => [r.array, r.plainObject, r.model],
+        () => {}
+      )
+    )
+
+    expect(findChildren(r, (node) => isArray(node)).has(r.array)).toBeTruthy()
+    expect(findChildren(r, (node) => isPlainObject(node)).has(r.plainObject)).toBeTruthy()
+    expect(findChildren(r, (node) => node instanceof M).has(r.model)).toBeTruthy()
+  })
+
+  test.each([false, true])("computed tree supports getChildrenObjects (deep=%j)", (deep) => {
+    const r = new R({})
+
+    // no children objects because the computed trees are not observed at this point
+    expect(getChildrenObjects(r, { deep }).size).toBe(0)
+
+    // cached when observed
+    autoDispose(
+      reaction(
+        () => [r.array, r.plainObject, r.model],
+        () => {}
+      )
+    )
+
+    const children = getChildrenObjects(r, { deep })
+    expect(children.size).toBe(3)
+    expect(children.has(r.array)).toBeTruthy()
+    expect(children.has(r.plainObject)).toBeTruthy()
+    expect(children.has(r.model)).toBeTruthy()
```

Comment:
I'm not sure if it is good that computed tree nodes appear in getChildrenObjects (not to mention having them magically appear/disappear when they are observed/unobserved is confusing)

Reply:
perhaps isComputedTreeNode could be used to filter them out?

Reply:
or at least they should be hidden under an optional parameter which is false (so they are hidden) by default

Reply:
This is somewhat related to [your comment about `walkTree`](https://github.com/xaviergonz/mobx-keystone/pull/383#discussion_r799012501). Whatever we decide there, I think `walkTree` and `getChildrenObjects` (and `findChildren`) should be consistent.

> [...] (not to mention having them magically appear/disappear when they are observed/unobserved is confusing)

I'm not sure whether I find it confusing or not. Isn't it similar to having a node attached to (= observed) or detached from (i.e. set to `undefined`; = unobserved) a model prop?

As mentioned in my `walkTree` comment:

> Also, my view on computed trees is that they are essentially equivalent to regular trees except they are immutable and, thus, action middlewares are useless because there are no action calls in computed trees. With that in mind, I think they should be treated no different than regular trees as much as possible, so e.g. tree traversal functions should behave the same for computed and regular trees.

From that perspective, I think it makes sense to me the way it is now. But I'm happy to be convinced otherwise. :slightly_smiling_face:

That said, would it make sense to set [`keepAlive: true`](https://mobx.js.org/computeds.html#keepalive) on the computed tree? This way, the computed property would always be observed and the behavior would be more consistent with regular tree nodes.

Reply:
Oh, and if `getChildrenObjects` filtered out computed tree nodes, then it would be impossible to call it on a computed tree node as the subtree root, too, so it would be entirely useless in computed trees.

Reply:
Maybe it could use an option to get those, like walkTree?

Reply:
> That said, would it make sense to set [keepAlive: true](https://mobx.js.org/computeds.html#keepalive) on the computed tree? This way, the computed property would always be observed and the behavior would be more consistent with regular tree nodes.

I think keepAlive would never ever release it (not 100% sure though), but yeah, perhaps there could be a "custom keep alive", a WeakMap<nonComputedObject, Set<ComputedObject>> that would be auto populated when the nonComputedObject is constructed (right before onInit?). If that were the case I'd have no problem keeping those nodes in walkTree/getChildrenObjects (and would be even preferable so there's no filtering overhead)

Reply:
https://medium.com/terria/when-and-why-does-mobxs-keepalive-cause-a-memory-leak-8c29feb9ff55#:~:text=Why%20use%20keepAlive%20%3F,computed%20value%20is%20automatically%20cached.&text=When%20one%20of%20the%20observables,and%20re%2Drun%20the%20derivation.

Reply:
Thanks for digging into the `keepAlive` topic! I agree, it looks like `keepAlive: true` should be fine. From what I could tell after settings this option, it is not immediate, so I added another late initialization hook just before `onInit`/`onLazyInit` and registered a callback that access the property ones. See ce5e5d1e4f96371eb8d6d5b761a68f51286623d7.

It seems to work fine after a couple of changes because eager evaluation caused infinite recursion with the previous test setup. The only thing I'm not able to figure out right now is why the `onChildAttachedTo` test fails (it's being skipped now). The callback is not called after `r.setValue(10)`, but `r.model.value` is `10` then. Perhaps this change is not considered a detachment of the old model instance and attachment of the new model instance?

Reply:
It happened because the old value was never getting detached first. I worked around it by using a cache of "tweaked values" and untweaking old values / tweaking new values when the inner computed value changed

Reply:
https://github.com/xaviergonz/mobx-keystone/pull/383/commits/d58cda24bcdfbf0be6a79146652a4e55c4e712fe

Reply:
Very nice! Thank you very much for fixing it! :pray:

I only made a minor change: I simplified `getComputedTreeNodeInfo`  by using the `getOrCreate` utility function (see 07a2176db75f06168382c43020a1ec6d92a4d41d).